base/multidimensional.jl

```diff
@@ -343,45 +368,75 @@ module IteratorsMD
     eltype(::Type{CartesianIndices{N,TT}}) where {N,TT} = CartesianIndex{N}
     IteratorSize(::Type{<:CartesianIndices{N}}) where {N} = Base.HasShape{N}()
 
-    @inline function iterate(iter::CartesianIndices)
-        iterfirst, iterlast = first(iter), last(iter)
-        if any(map(>, iterfirst.I, iterlast.I))
+    @propagate_inbounds function iterate(iter::CartesianIndices)
+        iterfirst = first(iter)
+        if !all(map(in, iterfirst.I, iter.indices))
             return nothing
         end
         iterfirst, iterfirst
     end
-    @inline function iterate(iter::CartesianIndices, state)
-        valid, I = __inc(state.I, first(iter).I, last(iter).I)
+    @propagate_inbounds function iterate(iter::CartesianIndices, state)
+        valid, I = __inc(state.I, iter.indices)
         valid || return nothing
         return CartesianIndex(I...), CartesianIndex(I...)
     end
 
     # increment & carry
-    @inline function inc(state, start, stop)
-        _, I = __inc(state, start, stop)
+    @propagate_inbounds function inc(state, indices)
+        _, I = __inc(state, indices)
         return CartesianIndex(I...)
     end
 
-    # increment post check to avoid integer overflow
-    @inline __inc(::Tuple{}, ::Tuple{}, ::Tuple{}) = false, ()
-    @inline function __inc(state::Tuple{Int}, start::Tuple{Int}, stop::Tuple{Int})
-        valid = state[1] < stop[1]
-        return valid, (state[1]+1,)
+    __inc(::Tuple{}, ::Tuple{}) = false, ()
+    @propagate_inbounds function __inc(state::Tuple{Int}, indices::Tuple{<:OrdinalRange})
+        rng = indices[1]
+        I = state[1] + step(rng)
+        valid = __isvalid_inc_range(I, rng) && __has_inc_overflow(state[1], rng)
```

Comment:
Wondering if we should use use the iteration of the underlying ranges? I'm no longer remembering why it was implemented this way.

Reply:
Yes, this is a better idea.

There are some uncaught StepRange overflow behavior #37876 and breaks some of the existing tests, I'll mark them as broken.

Reply:
> I'm no longer remembering why it was implemented this way.

Here's my guess. `iterate` protocol returns `Union{Nothing, Tuple}`, the current implementation eliminates this type instability by passing an explicit `valid` flag, which makes the whole `__inc` thing is type stable.

The version I tried locally is

```julia
    iterate(iter::CartesianIndices, state) = _maybe_cartesian(__iterate(iter.indices, state.I))
    _maybe_cartesian(::Nothing) = nothing
    _maybe_cartesian(I::Tuple) = CartesianIndex(I), CartesianIndex(I)

    __iterate(indices::Tuple{<:OrdinalRange}, state::Tuple{Int}) = _maybe_int(iterate(indices[1], state[1]))
    _maybe_int(::Nothing) = nothing
    _maybe_int(I::Tuple) = (I[1], )

    @propagate_inbounds function __iterate(indices, state)
        rng = indices[1]
        first_idx = _iterate_range(rng, state[1])
        if !isnothing(first_idx)
            # the first column still has elements
            return (first_idx, tail(state)...)
        end
        # otherwise, recursively iterate on other columns
        first_idx = first(rng)
        tailing_idxs = __iterate(tail(indices), tail(state))
        if !isnothing(tailing_idxs)
            return (first_idx, tailing_idxs...)
        end
        return nothing
    end

    @propagate_inbounds _iterate_range(rng::OrdinalRange, state) = __iterate_result(iterate(rng, state))
    __iterate_result(::Nothing) = nothing
    __iterate_result(t::Tuple) = first(t)
```

as you can see, three one-liners are defined to extract and bypass `nothing`, but this still shakes the entire types in `__iterate`.

As a comparison, 

```julia
R = CartesianIndices((0:1, 0:2))
i = CartesianIndex(1, 0)
```

```
julia> @btime iterate(R, i)
158.875 ns (4 allocations: 112 bytes) # my local version
22.060 ns (1 allocation: 48 bytes) # the current version
```